File: frontend/src/objects/battle/microservices/session/objects/command/BattleCommandModels.ts
import type { BattleCommandSkillIntents } from "../../../abilities/objects/skill/BattleCommandSkillIntents";
import type { SkillKind } from "../../../abilities/objects/skill/SkillKind";
import type { SkillOutcomeReason } from "../../../abilities/objects/skill/SkillOutcomeReason";
import type { SkillOutcomeStatus } from "../../../abilities/objects/skill/SkillOutcomeStatus";
import type { BattleWeaponSwitchDirection } from "../../../combat/objects/weapon/BattleWeaponSwitchDirection";
import type { BattleWeaponSwitchIndex } from "../../../combat/objects/weapon/BattleWeaponSwitchIndex";
import type {
  BattleId,
  BattleTick,
  ClientCommandSeq,
  EpochMillis,
  PlayerId,
  TicketId
} from "../../../../objects/core/BattleCoreScalars";
import type { BattleCommandReason } from "./BattleCommandReason";
import type { BattleCommandStatus } from "./BattleCommandStatus";

export interface BattleCommandVector {
  x: number;
  y: number;
}

export interface BattleCommandRequest {
  battleId: BattleId;
  playerId: PlayerId;
  ticketId: TicketId;
  clientTick: BattleTick;
  clientCommandSeq: ClientCommandSeq;
  movement: BattleCommandVector;
  aim: BattleCommandVector;
  primaryHeld: boolean;
  sprint: boolean;
  reloadPressed: boolean;
  skillIntents: BattleCommandSkillIntents;
  pointerWorld: BattleCommandVector | null;
  switchWeaponDirection: BattleWeaponSwitchDirection;
  switchWeaponIndex: BattleWeaponSwitchIndex | null;
}

export interface BattleCommandSkillOutcome {
  action: SkillKind;
  outcomeStatus: SkillOutcomeStatus;
  reason: SkillOutcomeReason | null;
}

export interface BattleCommandAccepted {
  battleId: BattleId;
  acceptedTick: BattleTick;
  acceptedCommandSeq: ClientCommandSeq;
  serverTime: EpochMillis;
  commandStatus: BattleCommandStatus;
  commandReason: BattleCommandReason | null;
  outcomes: BattleCommandSkillOutcome[];
}

File: backend/src/main/scala/services/battle/microservices/session/objects/command/BattleCommandModels.scala
package services.battle.microservices.session.objects.command

import _root_.services.battle.objects.core.{BattleId, BattleTick, ClientCommandSeq, EpochMillis, PlayerId}
import _root_.services.battle.microservices.queue.objects.queue.TicketId
import services.battle.microservices.abilities.objects.skill.{
  BattleCommandSkillIntents,
  SkillKind,
  SkillOutcomeReason,
  SkillOutcomeStatus
}
import services.battle.microservices.combat.objects.weapon.{BattleWeaponSwitchDirection, BattleWeaponSwitchIndex}

final case class BattleCommandVector(
  x: Double,
  y: Double
)

final case class BattleCommandRequest(
  battleId: BattleId,
  playerId: PlayerId,
  ticketId: TicketId,
  clientTick: BattleTick,
  clientCommandSeq: ClientCommandSeq,
  movement: BattleCommandVector,
  aim: BattleCommandVector,
  primaryHeld: Boolean,
  sprint: Boolean,
  reloadPressed: Boolean,
  skillIntents: BattleCommandSkillIntents,
  pointerWorld: Option[BattleCommandVector],
  switchWeaponDirection: BattleWeaponSwitchDirection,
  switchWeaponIndex: Option[BattleWeaponSwitchIndex]
)

final case class BattleCommandSkillOutcome(
  action: SkillKind,
  outcomeStatus: SkillOutcomeStatus,
  reason: Option[SkillOutcomeReason]
)

enum BattleCommandAcceptPath {
  case Fresh
  case Serialized
}

final case class BattleCommandServerDiagnostics(
  path: BattleCommandAcceptPath,
  receivedAt: EpochMillis,
  completedAt: EpochMillis,
  durationMs: Long,
  lockWaitMs: Long,
  lockHeldMs: Long,
  advanceMs: Long,
  commitRetryCount: Int,
  clientTick: BattleTick,
  acceptedTick: BattleTick,
  acceptedTickLag: Long,
  clientCommandSeq: ClientCommandSeq,
  acceptedCommandSeq: ClientCommandSeq,
  acceptedCommandSeqLag: Long
)

final case class BattleCommandAccepted(
  battleId: BattleId,
  acceptedTick: BattleTick,
  acceptedCommandSeq: ClientCommandSeq,
  serverTime: EpochMillis,
  commandStatus: BattleCommandStatus,
  commandReason: Option[BattleCommandReason],
  outcomes: Vector[BattleCommandSkillOutcome],
  serverDiagnostics: Option[BattleCommandServerDiagnostics] = None
)
